{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "00910a35-ffbc-4497-bcc8-21f008c326fb",
   "metadata": {},
   "source": [
    "# Grapher"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ebd1805a-55dc-409c-9ccf-512eb1f88667",
   "metadata": {},
   "source": [
    "## Dataset"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f7a8ec1d-73be-4b8c-95dc-79489e84998b",
   "metadata": {},
   "source": [
    "- `id`: This column represents the id of the variant in the following format: #chrom:pos:ref:alt (string).\n",
    "\n",
    "- `#chrom`: This column represents the chromosome number where the genetic variant is located.\n",
    "\n",
    "- `pos`: This is the position of the genetic variant on the chromosome.\n",
    "\n",
    "- `ref`: This column represents the reference allele (or variant) at the genomic position.\n",
    "\n",
    "- `alt`: This is the alternate allele observed at this position.\n",
    "\n",
    "- `rsids`: This stands for reference SNP cluster ID. It's a unique identifier for each variant used in the dbSNP database.\n",
    "\n",
    "- `nearest_genes`: This column represents the gene which is nearest to the variant.\n",
    "\n",
    "- `pval`: This represents the p-value, which is a statistical measure for the strength of evidence against the null hypothesis.\n",
    "\n",
    "- `mlogp`: This represents the minus log of the p-value, commonly used in genomic studies.\n",
    "\n",
    "- `beta`: The beta coefficient represents the effect size of the variant.\n",
    "\n",
    "- `sebeta`: This is the standard error of the beta coefficient.\n",
    "\n",
    "- `af_alt`: This is the allele frequency of the alternate variant in the general population.\n",
    "\n",
    "- `af_alt_cases`: This is the allele frequency of the alternate variant in the cases group.\n",
    "\n",
    "- `af_alt_controls`: This is the allele frequency of the alternate variant in the control group.\n",
    "\n",
    "- `causal`: This binary column indicates whether the variant is determined to be causal (1) or not (0).\n",
    "\n",
    "- `LD`: This binary column indicates whether the variant is determined to be in linkage disequilibrium (1) or not (0).\n",
    "\n",
    "- `lead`: This string column contains the id of the SNP of which the variant is in LD with (string).\n",
    "\n",
    "- `trait`: This column represents the trait associated with the variant. In this dataset, it is the response to the drug paracetamol and NSAIDs."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "113b033d-95f5-4584-b10c-72969a166612",
   "metadata": {},
   "source": [
    "## Load libraries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "d4345459-b1ef-477c-8f06-c8f54c194af0",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Python version: 3.11.3 (tags/v3.11.3:f3909b8, Apr  4 2023, 23:49:59) [MSC v.1934 64 bit (AMD64)]\n",
      "NumPy version: 1.24.3\n",
      "Pandas version: 2.0.1\n",
      "Matplotlib version: 3.7.1\n",
      "Scikit-learn version: 1.2.2\n",
      "Torch version: 2.0.0+cu118\n",
      "Torch Geometric version: 2.3.1\n",
      "NetworkX version: 3.0\n",
      "Using NVIDIA GeForce RTX 3060 Ti (cuda)\n",
      "CUDA version: 11.8\n",
      "Number of CUDA devices: 1\n"
     ]
    }
   ],
   "source": [
    "import sys\n",
    "import os\n",
    "import random\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import matplotlib\n",
    "import matplotlib.pyplot as plt\n",
    "import sklearn\n",
    "from sklearn.model_selection import train_test_split\n",
    "from sklearn.metrics import precision_recall_fscore_support, roc_auc_score, average_precision_score\n",
    "from sklearn.preprocessing import LabelEncoder, StandardScaler, OrdinalEncoder, OneHotEncoder\n",
    "from sklearn.pipeline import Pipeline\n",
    "from sklearn.compose import ColumnTransformer\n",
    "from sklearn.impute import SimpleImputer\n",
    "import torch\n",
    "import torch.nn.functional as F\n",
    "import torch_geometric\n",
    "from torch_geometric.data import Data\n",
    "from torch_geometric.nn import GCNConv, GATConv\n",
    "from torch_geometric.utils import to_undirected, negative_sampling\n",
    "import networkx as nx\n",
    "from scipy.spatial import cKDTree\n",
    "from typing import List, Dict\n",
    "import time\n",
    "\n",
    "# Print versions of imported libraries\n",
    "print(f\"Python version: {sys.version}\")\n",
    "print(f\"NumPy version: {np.__version__}\")\n",
    "print(f\"Pandas version: {pd.__version__}\")\n",
    "print(f\"Matplotlib version: {matplotlib.__version__}\")\n",
    "print(f\"Scikit-learn version: {sklearn.__version__}\")\n",
    "print(f\"Torch version: {torch.__version__}\")\n",
    "print(f\"Torch Geometric version: {torch_geometric.__version__}\")\n",
    "print(f\"NetworkX version: {nx.__version__}\")\n",
    "\n",
    "if torch.cuda.is_available():\n",
    "    device = torch.device(\"cuda\")          # Current CUDA device\n",
    "    print(f\"Using {torch.cuda.get_device_name()} ({device})\")\n",
    "    print(f\"CUDA version: {torch.version.cuda}\")\n",
    "    print(f\"Number of CUDA devices: {torch.cuda.device_count()}\")\n",
    "else:\n",
    "    print(\"CUDA is not available on this device.\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9ddfa8cb-01df-491f-94f0-c8c228256b4d",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Load data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "05fe5b06-7392-4d62-a1e8-af081cdf3e6c",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "dtypes = {\n",
    "    'id': 'string',\n",
    "    '#chrom': 'string',\n",
    "    'pos': 'int64',\n",
    "    'ref': 'string',\n",
    "    'alt': 'string',\n",
    "    'rsids': 'string',\n",
    "    'nearest_genes': 'string',\n",
    "    'pval': 'float64',\n",
    "    'mlogp': 'float64',\n",
    "    'beta': 'float64',\n",
    "    'sebeta': 'float64',\n",
    "    'af_alt': 'float64',\n",
    "    'af_alt_cases': 'float64',\n",
    "    'af_alt_controls': 'float64',\n",
    "    'causal': 'int64',\n",
    "    'LD': 'int64',\n",
    "    'lead': 'string',\n",
    "    'trait': 'string'\n",
    "}\n",
    "\n",
    "data = pd.read_csv('~/Desktop/gwas-graph/FinnGen/data/gwas-causal.csv', dtype=dtypes)\n",
    "\n",
    "# Assert column names\n",
    "expected_columns = ['#chrom', 'pos', 'ref', 'alt', 'rsids', 'nearest_genes', 'pval', 'mlogp', 'beta',\n",
    "                    'sebeta', 'af_alt', 'af_alt_cases', 'af_alt_controls', 'causal', 'LD', 'lead',\n",
    "                    'id', 'trait']\n",
    "assert set(data.columns) == set(expected_columns), \"Unexpected columns in the data DataFrame.\"\n",
    "\n",
    "# Assert data types\n",
    "expected_dtypes = {\n",
    "    'id': 'string',\n",
    "    '#chrom': 'string',\n",
    "    'pos': 'int64',\n",
    "    'ref': 'string',\n",
    "    'alt': 'string',\n",
    "    'rsids': 'string',\n",
    "    'nearest_genes': 'string',\n",
    "    'pval': 'float64',\n",
    "    'mlogp': 'float64',\n",
    "    'beta': 'float64',\n",
    "    'sebeta': 'float64',\n",
    "    'af_alt': 'float64',\n",
    "    'af_alt_cases': 'float64',\n",
    "    'af_alt_controls': 'float64',\n",
    "    'causal': 'int64',\n",
    "    'LD': 'int64',\n",
    "    'lead': 'string',\n",
    "    'trait': 'string'\n",
    "}\n",
    "\n",
    "for col, expected_dtype in expected_dtypes.items():\n",
    "    assert data[col].dtype == expected_dtype, f\"Unexpected data type for column {col}.\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "43d6d595-4ef8-41d5-b0a5-dbdc9f2d9b03",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total number of null values in each column:\n",
      "#chrom                    0\n",
      "pos                       0\n",
      "ref                       0\n",
      "alt                       0\n",
      "rsids               1366396\n",
      "nearest_genes        727855\n",
      "pval                      0\n",
      "mlogp                     0\n",
      "beta                      0\n",
      "sebeta                    0\n",
      "af_alt                    0\n",
      "af_alt_cases              0\n",
      "af_alt_controls           0\n",
      "causal                    0\n",
      "LD                        0\n",
      "id                        0\n",
      "lead               20168198\n",
      "trait                     0\n",
      "dtype: int64\n"
     ]
    }
   ],
   "source": [
    "# Check for total number of null values in each column\n",
    "null_counts = data.isnull().sum()\n",
    "\n",
    "print(\"Total number of null values in each column:\")\n",
    "print(null_counts)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "602a754a-7836-439e-801c-7efca6d2dfc0",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Data manipulation"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f2cc4255-7a42-4172-a91d-4df08665fccc",
   "metadata": {
    "tags": []
   },
   "source": [
    "### Create new rows per gene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "9b729972-b0d7-4e38-b8e2-aedd07506c98",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "data['nearest_genes'] = data['nearest_genes'].astype(str)\n",
    "\n",
    "# Assert column 'nearest_genes' is a string\n",
    "assert data['nearest_genes'].dtype == 'object', \"Column 'nearest_genes' is not of string type.\"\n",
    "\n",
    "# Split the gene names in the 'nearest_genes' column\n",
    "split_genes = data['nearest_genes'].str.split(',')\n",
    "\n",
    "# Flatten the list of split gene names\n",
    "flat_genes = [item for sublist in split_genes for item in sublist]\n",
    "\n",
    "# Create a new DataFrame by repeating rows and substituting the gene names\n",
    "data_new = data.loc[data.index.repeat(split_genes.str.len())].copy()\n",
    "data_new['nearest_genes'] = flat_genes\n",
    "\n",
    "# Assert the shape of the new DataFrame is as expected\n",
    "expected_shape = (len(flat_genes), data.shape[1])\n",
    "assert data_new.shape == expected_shape, \"Shape of the new DataFrame is not as expected.\"\n",
    "\n",
    "# Reset index to have a standard index\n",
    "data = data_new.reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "513cafa0-f3f8-47b5-9d22-96062449edde",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#chrom</th>\n",
       "      <th>pos</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>rsids</th>\n",
       "      <th>nearest_genes</th>\n",
       "      <th>pval</th>\n",
       "      <th>mlogp</th>\n",
       "      <th>beta</th>\n",
       "      <th>sebeta</th>\n",
       "      <th>af_alt</th>\n",
       "      <th>af_alt_cases</th>\n",
       "      <th>af_alt_controls</th>\n",
       "      <th>causal</th>\n",
       "      <th>LD</th>\n",
       "      <th>id</th>\n",
       "      <th>lead</th>\n",
       "      <th>trait</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>13668</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>rs2691328</td>\n",
       "      <td>OR4F5</td>\n",
       "      <td>0.944365</td>\n",
       "      <td>0.024860</td>\n",
       "      <td>-0.005926</td>\n",
       "      <td>0.084918</td>\n",
       "      <td>0.005842</td>\n",
       "      <td>0.005729</td>\n",
       "      <td>0.005863</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1:13668:G:A</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>T2D</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>14773</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>rs878915777</td>\n",
       "      <td>OR4F5</td>\n",
       "      <td>0.844305</td>\n",
       "      <td>0.073501</td>\n",
       "      <td>0.010088</td>\n",
       "      <td>0.051369</td>\n",
       "      <td>0.013495</td>\n",
       "      <td>0.013547</td>\n",
       "      <td>0.013485</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1:14773:C:T</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>T2D</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>15585</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>rs533630043</td>\n",
       "      <td>OR4F5</td>\n",
       "      <td>0.841908</td>\n",
       "      <td>0.074735</td>\n",
       "      <td>0.031464</td>\n",
       "      <td>0.157751</td>\n",
       "      <td>0.001113</td>\n",
       "      <td>0.001125</td>\n",
       "      <td>0.001110</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1:15585:G:A</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>T2D</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>16549</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>rs1262014613</td>\n",
       "      <td>OR4F5</td>\n",
       "      <td>0.343308</td>\n",
       "      <td>0.464316</td>\n",
       "      <td>0.241377</td>\n",
       "      <td>0.254711</td>\n",
       "      <td>0.000561</td>\n",
       "      <td>0.000620</td>\n",
       "      <td>0.000550</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1:16549:T:C</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>T2D</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>16567</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>rs1194064194</td>\n",
       "      <td>OR4F5</td>\n",
       "      <td>0.129883</td>\n",
       "      <td>0.886447</td>\n",
       "      <td>0.130736</td>\n",
       "      <td>0.086319</td>\n",
       "      <td>0.004170</td>\n",
       "      <td>0.004250</td>\n",
       "      <td>0.004154</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1:16567:G:C</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>T2D</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20565622</th>\n",
       "      <td>23</td>\n",
       "      <td>155697920</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>0.027115</td>\n",
       "      <td>1.566790</td>\n",
       "      <td>-0.013475</td>\n",
       "      <td>0.006097</td>\n",
       "      <td>0.290961</td>\n",
       "      <td>0.286054</td>\n",
       "      <td>0.291879</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>23:155697920:G:A</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>T2D</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20565623</th>\n",
       "      <td>23</td>\n",
       "      <td>155698443</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>0.178417</td>\n",
       "      <td>0.748564</td>\n",
       "      <td>-0.069907</td>\n",
       "      <td>0.051951</td>\n",
       "      <td>0.003259</td>\n",
       "      <td>0.003022</td>\n",
       "      <td>0.003304</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>23:155698443:C:A</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>T2D</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20565624</th>\n",
       "      <td>23</td>\n",
       "      <td>155698490</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>0.279640</td>\n",
       "      <td>0.553400</td>\n",
       "      <td>-0.020245</td>\n",
       "      <td>0.018725</td>\n",
       "      <td>0.024406</td>\n",
       "      <td>0.024312</td>\n",
       "      <td>0.024423</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>23:155698490:C:T</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>T2D</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20565625</th>\n",
       "      <td>23</td>\n",
       "      <td>155699751</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>0.078864</td>\n",
       "      <td>1.103120</td>\n",
       "      <td>-0.011284</td>\n",
       "      <td>0.006421</td>\n",
       "      <td>0.244829</td>\n",
       "      <td>0.241257</td>\n",
       "      <td>0.245498</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>23:155699751:C:T</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>T2D</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20565626</th>\n",
       "      <td>23</td>\n",
       "      <td>155700291</td>\n",
       "      <td>CAA</td>\n",
       "      <td>C</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>0.071590</td>\n",
       "      <td>1.145150</td>\n",
       "      <td>-0.011563</td>\n",
       "      <td>0.006418</td>\n",
       "      <td>0.245328</td>\n",
       "      <td>0.241724</td>\n",
       "      <td>0.246003</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>23:155700291:CAA:C</td>\n",
       "      <td>&lt;NA&gt;</td>\n",
       "      <td>T2D</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>20565627 rows × 18 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         #chrom        pos  ref alt         rsids nearest_genes      pval   \n",
       "0             1      13668    G   A     rs2691328         OR4F5  0.944365  \\\n",
       "1             1      14773    C   T   rs878915777         OR4F5  0.844305   \n",
       "2             1      15585    G   A   rs533630043         OR4F5  0.841908   \n",
       "3             1      16549    T   C  rs1262014613         OR4F5  0.343308   \n",
       "4             1      16567    G   C  rs1194064194         OR4F5  0.129883   \n",
       "...         ...        ...  ...  ..           ...           ...       ...   \n",
       "20565622     23  155697920    G   A          <NA>          <NA>  0.027115   \n",
       "20565623     23  155698443    C   A          <NA>          <NA>  0.178417   \n",
       "20565624     23  155698490    C   T          <NA>          <NA>  0.279640   \n",
       "20565625     23  155699751    C   T          <NA>          <NA>  0.078864   \n",
       "20565626     23  155700291  CAA   C          <NA>          <NA>  0.071590   \n",
       "\n",
       "             mlogp      beta    sebeta    af_alt  af_alt_cases   \n",
       "0         0.024860 -0.005926  0.084918  0.005842      0.005729  \\\n",
       "1         0.073501  0.010088  0.051369  0.013495      0.013547   \n",
       "2         0.074735  0.031464  0.157751  0.001113      0.001125   \n",
       "3         0.464316  0.241377  0.254711  0.000561      0.000620   \n",
       "4         0.886447  0.130736  0.086319  0.004170      0.004250   \n",
       "...            ...       ...       ...       ...           ...   \n",
       "20565622  1.566790 -0.013475  0.006097  0.290961      0.286054   \n",
       "20565623  0.748564 -0.069907  0.051951  0.003259      0.003022   \n",
       "20565624  0.553400 -0.020245  0.018725  0.024406      0.024312   \n",
       "20565625  1.103120 -0.011284  0.006421  0.244829      0.241257   \n",
       "20565626  1.145150 -0.011563  0.006418  0.245328      0.241724   \n",
       "\n",
       "          af_alt_controls  causal  LD                  id  lead trait  \n",
       "0                0.005863       0   0         1:13668:G:A  <NA>   T2D  \n",
       "1                0.013485       0   0         1:14773:C:T  <NA>   T2D  \n",
       "2                0.001110       0   0         1:15585:G:A  <NA>   T2D  \n",
       "3                0.000550       0   0         1:16549:T:C  <NA>   T2D  \n",
       "4                0.004154       0   0         1:16567:G:C  <NA>   T2D  \n",
       "...                   ...     ...  ..                 ...   ...   ...  \n",
       "20565622         0.291879       0   0    23:155697920:G:A  <NA>   T2D  \n",
       "20565623         0.003304       0   0    23:155698443:C:A  <NA>   T2D  \n",
       "20565624         0.024423       0   0    23:155698490:C:T  <NA>   T2D  \n",
       "20565625         0.245498       0   0    23:155699751:C:T  <NA>   T2D  \n",
       "20565626         0.246003       0   0  23:155700291:CAA:C  <NA>   T2D  \n",
       "\n",
       "[20565627 rows x 18 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "66eddaed-2f48-4440-b6d6-0dc247f23ea8",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Spec"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8fe89142-2b65-469b-b330-1620d303312b",
   "metadata": {},
   "source": [
    "`data` df:\n",
    "\n",
    "- `id`: This column represents the id of the variant in the following format: #chrom:pos:ref:alt (string).\n",
    "\n",
    "- `#chrom`: This column represents the chromosome number where the genetic variant is located.\n",
    "\n",
    "- `pos`: This is the position of the genetic variant on the chromosome.\n",
    "\n",
    "- `ref`: This column represents the reference allele (or variant) at the genomic position.\n",
    "\n",
    "- `alt`: This is the alternate allele observed at this position.\n",
    "\n",
    "- `rsids`: This stands for reference SNP cluster ID. It's a unique identifier for each variant used in the dbSNP database.\n",
    "\n",
    "- `nearest_genes`: This column represents the gene which is nearest to the variant.\n",
    "\n",
    "- `pval`: This represents the p-value, which is a statistical measure for the strength of evidence against the null hypothesis.\n",
    "\n",
    "- `mlogp`: This represents the minus log of the p-value, commonly used in genomic studies.\n",
    "\n",
    "- `beta`: The beta coefficient represents the effect size of the variant.\n",
    "\n",
    "- `sebeta`: This is the standard error of the beta coefficient.\n",
    "\n",
    "- `af_alt`: This is the allele frequency of the alternate variant in the general population.\n",
    "\n",
    "- `af_alt_cases`: This is the allele frequency of the alternate variant in the cases group.\n",
    "\n",
    "- `af_alt_controls`: This is the allele frequency of the alternate variant in the control group.\n",
    "\n",
    "- `causal`: This binary column indicates whether the variant is determined to be causal (1) or not (0).\n",
    "\n",
    "- `LD`: This binary column indicates whether the variant is determined to be in linkage disequilibrium (1) or not (0).\n",
    "\n",
    "- `lead`: This string column contains the id of the SNP of which the variant is in LD with (string).\n",
    "\n",
    "- `trait`: This column represents the trait associated with the variant. In this dataset, it is the response to the drug paracetamol and NSAIDs.\n",
    "\n",
    "**Task Overview**\n",
    "- The objective is to design and implement a link prediction deep neural network model for predicting Linkage Disequilibrium between SNPs.\n",
    "\n",
    "**Nodes and Their Features**\n",
    "- There is one types of node: SNP nodes.\n",
    "- *SNP Nodes*: Each SNP Node is characterized by various features, including `id`, `nearest_genes`, `#chrom`, `pos`, `ref`, `alt`, `beta`, `sebeta`, `af_alt`, and `af_alt_cases` columns.\n",
    "\n",
    "**Edges, Their Features, and Labels**\n",
    "- Edges represent relationships between nodes. There is one type of edge: SNP-SNP.\n",
    "- *SNP-SNP Edges*:\n",
    "  - These edges are undirected, linking an SNP Node (as identified by the `id` column) to another SNP Node (as identified by the `lead` column) in the same data row:\n",
    "    - An edge is created when `data['LD'] = 1`, signifying that the two SNPs are in linkage disequilibrium."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "71abe533-6610-4f82-b8d6-533eb010858d",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Graph creation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "4f0a2540-a59c-4869-a2f2-f2f93c3e4ae8",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of unique SNPs: 20170006\n",
      "Number of positive SNP-SNP edges: 1808\n",
      "Number of nodes: 20565627\n",
      "Number of edges: 1808\n",
      "Node feature dimension: 9\n",
      "Execution time: 129.99298095703125 seconds\n"
     ]
    }
   ],
   "source": [
    "def get_unique_snps(data: pd.DataFrame) -> Dict:\n",
    "    \"\"\"\n",
    "    Function to create mappings for SNPs to integer indices.\n",
    "    \"\"\"\n",
    "    assert 'id' in data.columns, \"id column is missing in the data DataFrame.\"\n",
    "    assert not data['id'].isnull().any(), \"id column contains NaN values.\"\n",
    "\n",
    "    return {snp: idx for idx, snp in enumerate(data['id'].unique())}\n",
    "\n",
    "\n",
    "def preprocess_snp_features(data: pd.DataFrame, snp_to_idx: Dict) -> pd.DataFrame:\n",
    "    \"\"\"\n",
    "    Function to create node feature vectors for SNPs and preprocess categorical and numerical features.\n",
    "    \"\"\"\n",
    "    assert 'id' in data.columns, \"id column is missing in the data DataFrame.\"\n",
    "    assert not data['id'].isnull().any(), \"id column contains NaN values.\"\n",
    "\n",
    "    snp_features = data.loc[data['id'].isin(snp_to_idx.keys()),\n",
    "                            ['id', 'nearest_genes', '#chrom', 'pos', 'ref', 'alt', 'beta', 'sebeta',\n",
    "                             'af_alt', 'af_alt_cases']].drop_duplicates().set_index('id').sort_index()\n",
    "\n",
    "    # Fill NaNs with appropriate replacements\n",
    "    nan_replacements = {'nearest_genes': 'N/A', '#chrom': 'N/A', 'pos': 0, 'ref': 'N/A', 'alt': 'N/A',\n",
    "                        'beta': 0, 'sebeta': 0, 'af_alt': 0, 'af_alt_cases': 0}\n",
    "\n",
    "    for col in snp_features.columns:\n",
    "        assert col in nan_replacements, f\"{col} column is missing in nan_replacements dictionary.\"\n",
    "\n",
    "    # Impute missing values with appropriate replacements\n",
    "    imputer = SimpleImputer(missing_values=np.nan, strategy='constant', fill_value=nan_replacements)\n",
    "    snp_features = pd.DataFrame(imputer.fit_transform(snp_features), columns=snp_features.columns)\n",
    "\n",
    "    # Assert no empty or null values in features\n",
    "    assert not snp_features.isnull().any().any(), \"Features contain null values.\"\n",
    "    assert not snp_features.isna().any().any(), \"Features contain NaN values.\"\n",
    "    assert snp_features.size > 0, \"Features are empty.\"\n",
    "\n",
    "    # Label encoding for categorical columns and standardize numerical features\n",
    "    le = LabelEncoder()\n",
    "    scaler = StandardScaler()\n",
    "    for col in snp_features.columns:\n",
    "        if col in ['ref', 'alt', 'nearest_genes', '#chrom']:\n",
    "            snp_features[col] = le.fit_transform(snp_features[col].astype(str))\n",
    "        else:\n",
    "            snp_features[col] = pd.Series(scaler.fit_transform(snp_features[col].to_frame()).flatten())\n",
    "\n",
    "    return snp_features\n",
    "\n",
    "\n",
    "def preprocess_positive_edges(data: pd.DataFrame, snp_to_idx: Dict) -> torch.Tensor:\n",
    "    \"\"\"\n",
    "    Function to create positive SNP-SNP edges and preprocess them.\n",
    "    \"\"\"\n",
    "    assert 'id' in data.columns, \"id column is missing in the data DataFrame.\"\n",
    "    assert 'lead' in data.columns, \"lead column is missing in the data DataFrame.\"\n",
    "    assert not data['id'].isnull().any(), \"id column contains NaN values.\"\n",
    "\n",
    "    positive_edges_snp_snp = data.loc[(data['LD'] == 1) & (data['id'].isin(snp_to_idx)) & (data['lead'].isin(snp_to_idx)),\n",
    "                                      ['id', 'lead']].drop_duplicates().applymap(snp_to_idx.get).values\n",
    "    assert positive_edges_snp_snp.size > 0, \"No positive SNP-SNP edges found.\"\n",
    "\n",
    "    return torch.tensor(positive_edges_snp_snp, dtype=torch.long).t().contiguous()\n",
    "\n",
    "\n",
    "def create_pytorch_graph(features: torch.Tensor, edges: torch.Tensor) -> Data:\n",
    "    \"\"\"\n",
    "    Function to create the PyTorch Geometric graph.\n",
    "    \"\"\"\n",
    "    assert isinstance(features, torch.Tensor), \"features must be a torch.Tensor.\"\n",
    "    assert isinstance(edges, torch.Tensor), \"edges must be a torch.Tensor.\"\n",
    "\n",
    "    # Create edge labels (+1 for positive edges)\n",
    "    edge_attr = torch.ones(edges.size(1), dtype=torch.float)\n",
    "\n",
    "    return Data(x=features, edge_index=edges, edge_attr=edge_attr)\n",
    "\n",
    "\n",
    "assert 'data' in globals(), \"data variable is not defined.\"\n",
    "\n",
    "start_time = time.time()\n",
    "\n",
    "snp_to_idx = get_unique_snps(data)\n",
    "assert len(snp_to_idx) > 0, \"No unique SNPs found.\"\n",
    "print(f\"Number of unique SNPs: {len(snp_to_idx)}\")\n",
    "\n",
    "snp_features = preprocess_snp_features(data, snp_to_idx)\n",
    "features = torch.tensor(snp_features.values, dtype=torch.float)\n",
    "\n",
    "positive_edges_snp_snp = preprocess_positive_edges(data, snp_to_idx)\n",
    "assert positive_edges_snp_snp.size(1) > 0, \"No positive SNP-SNP edges found.\"\n",
    "print(f\"Number of positive SNP-SNP edges: {positive_edges_snp_snp.size(1)}\")\n",
    "\n",
    "graph = create_pytorch_graph(features, positive_edges_snp_snp)\n",
    "\n",
    "print(f\"Number of nodes: {graph.num_nodes}\")\n",
    "print(f\"Number of edges: {graph.num_edges}\")\n",
    "print(f\"Node feature dimension: {graph.num_node_features}\")\n",
    "\n",
    "elapsed_time = time.time() - start_time\n",
    "print(f\"Execution time: {elapsed_time} seconds\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a60cae52-2493-43ae-9e27-e7dc67c3a7f9",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Graph stats"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "76c98f83-a35d-4010-8adf-078fd4fbe799",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Graph stats:\n",
      "Number of nodes: 20565627\n",
      "Number of positive SNP-SNP edges: 1808\n",
      "Number of edges: 1808\n",
      "Node feature dimension: 9\n",
      "Density: 0.0000000000\n",
      "Features with NaN values:\n",
      "[]\n"
     ]
    }
   ],
   "source": [
    "from torch_geometric.utils import degree\n",
    "\n",
    "def print_graph_stats(graph, positive_edges_snp_snp):\n",
    "    print(f\"Number of nodes: {graph.num_nodes}\")\n",
    "    print(f\"Number of positive SNP-SNP edges: {positive_edges_snp_snp.size(1)}\")\n",
    "    print(f\"Number of edges: {graph.num_edges}\")\n",
    "    print(f\"Node feature dimension: {graph.num_node_features}\")\n",
    "\n",
    "    # Compute and print degree-related stats\n",
    "    degrees = degree(graph.edge_index[0].long(), num_nodes=graph.num_nodes)\n",
    "    average_degree = degrees.float().mean().item()\n",
    "    median_degree = np.median(degrees.numpy())\n",
    "    std_degree = degrees.float().std().item()\n",
    "\n",
    "    # Assert average degree, median degree, and std degree\n",
    "    assert isinstance(average_degree, float), \"Average degree is not a float.\"\n",
    "    assert isinstance(std_degree, float), \"Standard deviation of degree is not a float.\"\n",
    "\n",
    "    # Density is the ratio of actual edges to the maximum number of possible edges\n",
    "    num_possible_edges = graph.num_nodes * (graph.num_nodes - 1) / 2\n",
    "    density = graph.num_edges / num_possible_edges\n",
    "\n",
    "    # Assert density\n",
    "    assert isinstance(density, float), \"Density is not a float.\"\n",
    "\n",
    "    print(f\"Density: {density:.10f}\")\n",
    "\n",
    "    # Check for NaN values in features\n",
    "    nan_mask = torch.isnan(graph.x)\n",
    "    nan_features = []\n",
    "    for feature_idx, feature_name in enumerate(snp_features.columns):\n",
    "        if nan_mask[:, feature_idx].any():\n",
    "            nan_features.append(feature_name)\n",
    "\n",
    "    print(\"Features with NaN values:\")\n",
    "    print(nan_features)\n",
    "\n",
    "# Print graph stats\n",
    "print(\"Graph stats:\")\n",
    "print_graph_stats(graph, positive_edges_snp_snp)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4a51e96d-660c-4b4f-8ac4-9a07dba5bcc1",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Data splitting"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "8a5650ea-db41-4f17-8599-86b148482867",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Data(x=[20565627, 9], edge_index=[2, 788], edge_attr=[788], pos_edge_label=[394], pos_edge_label_index=[2, 394], neg_edge_label=[394000], neg_edge_label_index=[2, 394000])\n",
      "Data(x=[20565627, 9], edge_index=[2, 788], edge_attr=[788], pos_edge_label=[196], pos_edge_label_index=[2, 196], neg_edge_label=[196000], neg_edge_label_index=[2, 196000])\n",
      "Data(x=[20565627, 9], edge_index=[2, 1180], edge_attr=[1180], pos_edge_label=[196], pos_edge_label_index=[2, 196], neg_edge_label=[196000], neg_edge_label_index=[2, 196000])\n"
     ]
    }
   ],
   "source": [
    "from torch_geometric.transforms import RandomLinkSplit\n",
    "\n",
    "transform = RandomLinkSplit(neg_sampling_ratio=1000, num_val=0.25, num_test=0.25, is_undirected=True, split_labels=True)\n",
    "\n",
    "graph_train, graph_val, graph_test = transform(graph)\n",
    "\n",
    "# Assert graph_train\n",
    "assert isinstance(graph_train, Data), \"graph_train is not an instance of torch_geometric.data.Data.\"\n",
    "assert graph_train.num_nodes == graph.num_nodes, \"Number of nodes in graph_train does not match the original graph.\"\n",
    "\n",
    "# Assert graph_val\n",
    "assert isinstance(graph_val, Data), \"graph_val is not an instance of torch_geometric.data.Data.\"\n",
    "\n",
    "# Assert graph_test\n",
    "assert isinstance(graph_test, Data), \"graph_test is not an instance of torch_geometric.data.Data.\"\n",
    "\n",
    "print(graph_train)\n",
    "print(graph_val)\n",
    "print(graph_test)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8665050a-5cd9-4960-a2ef-8a4cf2cb5e40",
   "metadata": {},
   "source": [
    "## Models"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "08d2557f-d84e-4a1b-bda8-7b26620f526b",
   "metadata": {},
   "source": [
    "### GCN"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "8ae175ab-f1da-4ac9-bd1c-f180eb8e2313",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Epoch: 1, Loss: 17929294.0000, Val ROC-AUC: 0.5370016920, Val Pos Edge Acc: 0.7406369141, Gamma: 2.00\n",
      "Epoch: 2, Loss: 291161.2188, Val ROC-AUC: 0.5075635933, Val Pos Edge Acc: 0.3049705397, Gamma: 2.10\n",
      "Epoch: 3, Loss: 8829.8379, Val ROC-AUC: 0.5056755388, Val Pos Edge Acc: 0.0354033721, Gamma: 2.20\n",
      "Epoch: 4, Loss: 1.4206, Val ROC-AUC: 0.5000000000, Val Pos Edge Acc: 0.9980019980, Gamma: 2.30\n",
      "Epoch: 5, Loss: 0.1406, Val ROC-AUC: 0.5000000000, Val Pos Edge Acc: 0.9980019980, Gamma: 2.30\n",
      "Epoch: 6, Loss: 0.1406, Val ROC-AUC: 0.5000000000, Val Pos Edge Acc: 0.9980019980, Gamma: 2.30\n",
      "Epoch: 7, Loss: 0.1406, Val ROC-AUC: 0.5294697782, Val Pos Edge Acc: 0.1944993782, Gamma: 2.30\n",
      "Epoch: 8, Loss: 5.6884, Val ROC-AUC: 0.5000000000, Val Pos Edge Acc: 0.0019980020, Gamma: 2.40\n",
      "Epoch: 9, Loss: 0.1315, Val ROC-AUC: 0.5000000000, Val Pos Edge Acc: 0.9980019980, Gamma: 2.40\n",
      "Epoch: 10, Loss: 0.1312, Val ROC-AUC: 0.5000000000, Val Pos Edge Acc: 0.9980019980, Gamma: 2.40\n",
      "Epoch: 11, Loss: 0.1312, Val ROC-AUC: 0.5000000000, Val Pos Edge Acc: 0.9980019980, Gamma: 2.40\n",
      "Early stopping!\n"
     ]
    }
   ],
   "source": [
    "import torch.nn as nn\n",
    "from torch_geometric.nn import SAGEConv, GINConv\n",
    "\n",
    "torch.cuda.empty_cache()\n",
    "\n",
    "# Define the GCN model\n",
    "class GCN(torch.nn.Module):\n",
    "    def __init__(self, in_channels, hidden_channels):\n",
    "        super(GCN, self).__init__()\n",
    "        self.conv1 = GCNConv(in_channels, hidden_channels)\n",
    "        self.conv2 = GCNConv(hidden_channels, hidden_channels)\n",
    "\n",
    "    def forward(self, x, edge_index):\n",
    "        # First conv layer\n",
    "        x = self.conv1(x, edge_index)\n",
    "        x = F.relu(x)\n",
    "        x = F.dropout(x, p=0.75, training=self.training)\n",
    "\n",
    "        # Second conv layer\n",
    "        x = self.conv2(x, edge_index)\n",
    "        x = F.relu(x)\n",
    "\n",
    "        return x\n",
    "\n",
    "\n",
    "# Train and evaluate the model\n",
    "device = torch.device(\"cuda\" if torch.cuda.is_available() else \"cpu\")\n",
    "model = GCN(in_channels=9, hidden_channels=4).to(device)\n",
    "optimizer = torch.optim.Adam(model.parameters(), lr=0.1, weight_decay=5e-4)\n",
    "scheduler = torch.optim.lr_scheduler.StepLR(optimizer, step_size=30, gamma=0.1)\n",
    "\n",
    "graph_train = graph_train.to(device)\n",
    "graph_val = graph_val.to(device)\n",
    "graph_test = graph_test.to(device)\n",
    "\n",
    "\n",
    "from torch.nn import BCEWithLogitsLoss\n",
    "\n",
    "def compute_alpha_beta(targets):\n",
    "    num_positives = targets.sum()\n",
    "    num_samples = len(targets)\n",
    "    beta = num_positives / num_samples  # proportion of positive instances\n",
    "    alpha = 1 - beta  # proportion of negative instances\n",
    "    return alpha, beta\n",
    "\n",
    "def compute_adaptive_gamma(val_roc_auc, prev_val_roc_auc, gamma, increment_factor=0.1):\n",
    "    if val_roc_auc < prev_val_roc_auc:  # If the performance decreases, increase gamma\n",
    "        gamma += increment_factor\n",
    "    return gamma\n",
    "\n",
    "# Train function\n",
    "def train(prev_val_roc_auc, gamma):\n",
    "    model.train()\n",
    "    optimizer.zero_grad()\n",
    "    z = model(graph_train.x.float(), graph_train.edge_index)\n",
    "\n",
    "    # Only consider positive edges for the positive score calculation\n",
    "    pos_edge_index = graph_train.pos_edge_label_index\n",
    "    pos = (z[pos_edge_index[0]] * z[pos_edge_index[1]]).sum(dim=-1)\n",
    "\n",
    "    # Use the pre-generated negative labels\n",
    "    neg_edge_index = graph_train.neg_edge_label_index\n",
    "    neg = (z[neg_edge_index[0]] * z[neg_edge_index[1]]).sum(dim=-1)\n",
    "\n",
    "    logits = torch.cat([pos, neg], dim=0)\n",
    "    targets = torch.tensor([1] * pos.size(0) + [0] * neg.size(0), dtype=torch.float32).to(device)\n",
    "\n",
    "    # Compute dynamic alpha and beta\n",
    "    alpha, beta = compute_alpha_beta(targets)\n",
    "\n",
    "    # Apply Focal Loss\n",
    "    bce_loss = BCEWithLogitsLoss(reduction='none')\n",
    "    bce = bce_loss(logits, targets)\n",
    "    pt = torch.exp(-bce)  # Compute the focal term\n",
    "    focal_loss = (alpha * (1 - pt) ** gamma * bce).mean()  # Compute the focal loss\n",
    "\n",
    "    focal_loss.backward()\n",
    "    \n",
    "    torch.nn.utils.clip_grad_norm_(model.parameters(), 1.0)\n",
    "\n",
    "    optimizer.step()\n",
    "    return focal_loss.item()\n",
    "\n",
    "def pos_edge_accuracy(preds, true_labels):\n",
    "    # Get predictions and true labels for positive edges\n",
    "    pos_preds = preds[:len(true_labels)//2]\n",
    "    pos_labels = true_labels[:len(true_labels)//2]\n",
    "\n",
    "    # Convert predictions to binary labels\n",
    "    bin_preds = (pos_preds > 0.5).astype(int)\n",
    "\n",
    "    # Compute the accuracy\n",
    "    accuracy = (bin_preds == pos_labels).mean()\n",
    "\n",
    "    return accuracy\n",
    "\n",
    "\n",
    "# Evaluation function\n",
    "def evaluate(graph):\n",
    "    model.eval()\n",
    "    with torch.no_grad():\n",
    "        z = model(graph.x.float(), graph.edge_index)\n",
    "\n",
    "        pos_edge_index = graph.pos_edge_label_index\n",
    "        pos = torch.sigmoid((z[pos_edge_index[0]] * z[pos_edge_index[1]]).sum(dim=-1)).view(-1)\n",
    "\n",
    "        neg_edge_index = graph.neg_edge_label_index\n",
    "        neg = torch.sigmoid((z[neg_edge_index[0]] * z[neg_edge_index[1]]).sum(dim=-1)).view(-1)\n",
    "\n",
    "        preds = torch.cat([pos, neg]).cpu().numpy()\n",
    "        true_labels = np.concatenate([np.ones_like(pos.cpu().numpy()), np.zeros_like(neg.cpu().numpy())])\n",
    "\n",
    "        # Handle NaN values\n",
    "        preds = np.nan_to_num(preds, nan=0.5)\n",
    "\n",
    "        roc_auc = roc_auc_score(true_labels, preds)\n",
    "        pos_edge_acc = pos_edge_accuracy(preds, true_labels)\n",
    "\n",
    "        return roc_auc, pos_edge_acc\n",
    "\n",
    "\n",
    "gamma = 2.0  # Starting value for gamma\n",
    "prev_val_roc_auc = 0.0  # Initial performance value\n",
    "best_val_roc_auc = 0.0  # Best validation ROC-AUC score\n",
    "early_stop = 10  # Number of epochs to continue without improvement before stopping\n",
    "epochs_no_improve = 0  # Counter for epochs without improvement\n",
    "\n",
    "for epoch in range(150):\n",
    "    loss = train(prev_val_roc_auc, gamma)\n",
    "    torch.cuda.empty_cache()  # Release unnecessary GPU memory\n",
    "    scheduler.step()\n",
    "    with torch.no_grad():  # Use torch.no_grad() during training\n",
    "        val_roc_auc, val_pos_edge_acc = evaluate(graph_val)\n",
    "        gamma = compute_adaptive_gamma(val_roc_auc, prev_val_roc_auc, gamma)\n",
    "        prev_val_roc_auc = val_roc_auc\n",
    "    print(f\"Epoch: {epoch + 1}, Loss: {loss:.4f}, Val ROC-AUC: {val_roc_auc:.10f}, Val Pos Edge Acc: {val_pos_edge_acc:.10f}, Gamma: {gamma:.2f}\")\n",
    "    if val_roc_auc > best_val_roc_auc:\n",
    "        best_val_roc_auc = val_roc_auc\n",
    "        epochs_no_improve = 0\n",
    "    else:\n",
    "        epochs_no_improve += 1\n",
    "\n",
    "    if epochs_no_improve == early_stop:\n",
    "        print(\"Early stopping!\")\n",
    "        break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0bba9103-b4af-4311-9bd8-033e04770339",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
