{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "00910a35-ffbc-4497-bcc8-21f008c326fb",
   "metadata": {
    "tags": []
   },
   "source": [
    "# FinnGen - Data Engineering Notebook - Causal"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a6fd8be8-14cc-4d80-a944-76c5235c078b",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Data Descriptions"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c1846d70-3c09-4b34-a05b-b72706e828fa",
   "metadata": {
    "tags": []
   },
   "source": [
    "### gwas"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8535f83a-6468-41bb-b3dc-6e678783c019",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "\n",
    "The `{endpoint}.gz` file has the following structure:\n",
    "\n",
    "| Column name   | Description                                                 |\n",
    "| ------------- | ----------------------------------------------------------- |\n",
    "| #chrom        | chromosome on build GRCh38 (1-23)                           |\n",
    "| pos           | position in base pairs on build GRCh38                       |\n",
    "| ref           | reference allele                                            |\n",
    "| alt           | alternative allele (effect allele)                           |\n",
    "| rsids         | variant identifier                                          |\n",
    "| nearest_genes | nearest gene(s) (comma separated) from variant               |\n",
    "| pval          | p-value from [source]                                        |\n",
    "| mlogp         | -log10(p-value)                                             |\n",
    "| beta          | effect size (log(OR) scale) estimated with [source]          |\n",
    "| sebeta        | standard error of effect size estimated with [source]        |\n",
    "| af_alt        | alternative (effect) allele frequency                        |\n",
    "| af_alt_cases  | alternative (effect) allele frequency among cases            |\n",
    "| af_alt_controls | alternative (effect) allele frequency among controls         |\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7a7cea15-a49b-4ee2-87a5-bf625453328e",
   "metadata": {
    "tags": []
   },
   "source": [
    "### finemap"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "10f3aafa-e866-463a-8cd1-a7416b1ff129",
   "metadata": {},
   "source": [
    "{endpoint}.SUSIE.snp.bgz` contains variant summaries with credible set information and has the following structure:\n",
    "\n",
    "| Column name    | Description                                                        |\n",
    "| -------------- | ------------------------------------------------------------------ |\n",
    "| trait          | endpoint name                                                      |\n",
    "| region         | chr:start-end                                                      |\n",
    "| v              | variant identifier                                                 |\n",
    "| rsid           | rs variant identifier                                              |\n",
    "| chromosome     | chromosome on build GRCh38 (1-22, X)                                |\n",
    "| position       | position in base pairs on build GRCh38                              |\n",
    "| allele1        | reference allele                                                   |\n",
    "| allele2        | alternative allele (effect allele)                                  |\n",
    "| maf            | minor allele frequency                                             |\n",
    "| beta           | effect size GWAS                                                   |\n",
    "| se             | standard error GWAS                                                |\n",
    "| p              | p-value GWAS                                                       |\n",
    "| mean           | posterior expectation of true effect size                           |\n",
    "| sd             | posterior standard deviation of true effect size                   |\n",
    "| prob           | posterior probability of association                                |\n",
    "| cs             | identifier of 95% credible set (-1 = variant is not part of credible set) |\n",
    "| lead_r2        | r2 value to a lead variant (the one with maximum PIP) in a credible set |\n",
    "| alphax         | posterior inclusion probability for the x-th single effect (x := 1..L where L is the number of single effects (causal variants) specified; default: L = 10) |"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "018acd43-12a7-41be-bf26-fc1d03dd7ef7",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Libraries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "483d7bbb-33ef-4893-b03c-6173b7253544",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Python version: 3.11.4 (tags/v3.11.4:d2340ef, Jun  7 2023, 05:45:37) [MSC v.1934 64 bit (AMD64)]\n",
      "Pandas version: 1.5.3\n",
      "NumPy version: 1.24.1\n"
     ]
    }
   ],
   "source": [
    "import sys\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import requests\n",
    "import time\n",
    "from concurrent.futures import ThreadPoolExecutor\n",
    "\n",
    "\n",
    "\n",
    "print(\"Python version:\", sys.version)\n",
    "print(\"Pandas version:\", pd.__version__)\n",
    "print(\"NumPy version:\", np.__version__)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "445033a6-7000-4fb6-9c4d-cfe03f570b1c",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Import data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "f1c04950-d2ca-4421-be63-297d26fc3096",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "C:\\Users\\Windows\\Desktop\\GeoGWAS\\FinnGen\\notebooks\\precausal\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "\n",
    "# Get the current working directory\n",
    "current_directory = os.getcwd()\n",
    "\n",
    "print(current_directory)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "d4345459-b1ef-477c-8f06-c8f54c194af0",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Read the 'finemap' file into a pandas DataFrame\n",
    "finemap = pd.read_csv('C:/Users/Windows/Desktop/GeoGWAS/FinnGen/data/finemapping_summary_finngen_R9_I9_HYPTENS.SUSIE_99.snp.filter.tsv', low_memory=False, sep='\\t')\n",
    "\n",
    "\n",
    "# Read the 'causal' file into a pandas DataFrame\n",
    "precausal = pd.read_csv('C:/Users/Windows/Desktop/GeoGWAS/FinnGen/data/finemapping_summary_finngen_R9_I9_HYPTENS.SUSIE.cred.summary.tsv', low_memory=False, sep='\\t')\n",
    "\n",
    "\n",
    "# Read the 'causal' file into a pandas DataFrame\n",
    "causal = pd.read_csv('C:/Users/Windows/Desktop/GeoGWAS/FinnGen/data/causal-hyp.tsv', low_memory=False, sep='\\t')\n",
    "\n",
    "\n",
    "# Read the 'gwas' file into a pandas DataFrame\n",
    "gwas = pd.read_csv('C:/Users/Windows/Desktop/GeoGWAS/FinnGen/data/summary_stats_finngen_R9_I9_HYPTENS.tsv', low_memory=False, sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "ffc11568-26b1-4283-b6b6-e1fd478f29e2",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "NaNs and missing values in 'gwas':\n",
      "#chrom                   0\n",
      "pos                      0\n",
      "ref                      0\n",
      "alt                      0\n",
      "rsids              1366441\n",
      "nearest_genes       727861\n",
      "pval                     0\n",
      "mlogp                    0\n",
      "beta                     0\n",
      "sebeta                   0\n",
      "af_alt                   0\n",
      "af_alt_cases             0\n",
      "af_alt_controls          0\n",
      "dtype: int64\n"
     ]
    }
   ],
   "source": [
    "print(\"NaNs and missing values in 'gwas':\")\n",
    "empty = gwas.isna().sum()\n",
    "print(empty)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3610b12e-d085-49a4-91e8-273fcb9dad32",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Explore data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "e3742572-cf8b-48a5-a540-4b99d41324b1",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#chrom</th>\n",
       "      <th>pos</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>rsids</th>\n",
       "      <th>nearest_genes</th>\n",
       "      <th>pval</th>\n",
       "      <th>mlogp</th>\n",
       "      <th>beta</th>\n",
       "      <th>sebeta</th>\n",
       "      <th>af_alt</th>\n",
       "      <th>af_alt_cases</th>\n",
       "      <th>af_alt_controls</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>13668</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>rs2691328</td>\n",
       "      <td>OR4F5</td>\n",
       "      <td>0.106658</td>\n",
       "      <td>0.972006</td>\n",
       "      <td>-0.114822</td>\n",
       "      <td>0.071168</td>\n",
       "      <td>0.005846</td>\n",
       "      <td>0.005683</td>\n",
       "      <td>0.005914</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>14773</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>rs878915777</td>\n",
       "      <td>OR4F5</td>\n",
       "      <td>0.620115</td>\n",
       "      <td>0.207528</td>\n",
       "      <td>-0.021548</td>\n",
       "      <td>0.043470</td>\n",
       "      <td>0.013501</td>\n",
       "      <td>0.013448</td>\n",
       "      <td>0.013524</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>15585</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>rs533630043</td>\n",
       "      <td>OR4F5</td>\n",
       "      <td>0.859628</td>\n",
       "      <td>0.065689</td>\n",
       "      <td>-0.023716</td>\n",
       "      <td>0.134105</td>\n",
       "      <td>0.001112</td>\n",
       "      <td>0.001117</td>\n",
       "      <td>0.001109</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>16549</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>rs1262014613</td>\n",
       "      <td>OR4F5</td>\n",
       "      <td>0.321844</td>\n",
       "      <td>0.492355</td>\n",
       "      <td>-0.215787</td>\n",
       "      <td>0.217818</td>\n",
       "      <td>0.000563</td>\n",
       "      <td>0.000556</td>\n",
       "      <td>0.000566</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>16567</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>rs1194064194</td>\n",
       "      <td>OR4F5</td>\n",
       "      <td>0.764225</td>\n",
       "      <td>0.116779</td>\n",
       "      <td>0.021523</td>\n",
       "      <td>0.071757</td>\n",
       "      <td>0.004192</td>\n",
       "      <td>0.004207</td>\n",
       "      <td>0.004186</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20170229</th>\n",
       "      <td>23</td>\n",
       "      <td>155697920</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.435606</td>\n",
       "      <td>0.360906</td>\n",
       "      <td>0.004066</td>\n",
       "      <td>0.005215</td>\n",
       "      <td>0.291210</td>\n",
       "      <td>0.291159</td>\n",
       "      <td>0.291231</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20170230</th>\n",
       "      <td>23</td>\n",
       "      <td>155698443</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.559723</td>\n",
       "      <td>0.252027</td>\n",
       "      <td>0.025926</td>\n",
       "      <td>0.044450</td>\n",
       "      <td>0.003263</td>\n",
       "      <td>0.003298</td>\n",
       "      <td>0.003248</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20170231</th>\n",
       "      <td>23</td>\n",
       "      <td>155698490</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.007623</td>\n",
       "      <td>2.117900</td>\n",
       "      <td>-0.043135</td>\n",
       "      <td>0.016165</td>\n",
       "      <td>0.024340</td>\n",
       "      <td>0.023984</td>\n",
       "      <td>0.024489</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20170232</th>\n",
       "      <td>23</td>\n",
       "      <td>155699751</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.160090</td>\n",
       "      <td>0.795637</td>\n",
       "      <td>0.007715</td>\n",
       "      <td>0.005492</td>\n",
       "      <td>0.245151</td>\n",
       "      <td>0.245738</td>\n",
       "      <td>0.244904</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20170233</th>\n",
       "      <td>23</td>\n",
       "      <td>155700291</td>\n",
       "      <td>CAA</td>\n",
       "      <td>C</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.149323</td>\n",
       "      <td>0.825873</td>\n",
       "      <td>0.007914</td>\n",
       "      <td>0.005489</td>\n",
       "      <td>0.245648</td>\n",
       "      <td>0.246278</td>\n",
       "      <td>0.245383</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>20170234 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          #chrom        pos  ref alt         rsids nearest_genes      pval  \\\n",
       "0              1      13668    G   A     rs2691328         OR4F5  0.106658   \n",
       "1              1      14773    C   T   rs878915777         OR4F5  0.620115   \n",
       "2              1      15585    G   A   rs533630043         OR4F5  0.859628   \n",
       "3              1      16549    T   C  rs1262014613         OR4F5  0.321844   \n",
       "4              1      16567    G   C  rs1194064194         OR4F5  0.764225   \n",
       "...          ...        ...  ...  ..           ...           ...       ...   \n",
       "20170229      23  155697920    G   A           NaN           NaN  0.435606   \n",
       "20170230      23  155698443    C   A           NaN           NaN  0.559723   \n",
       "20170231      23  155698490    C   T           NaN           NaN  0.007623   \n",
       "20170232      23  155699751    C   T           NaN           NaN  0.160090   \n",
       "20170233      23  155700291  CAA   C           NaN           NaN  0.149323   \n",
       "\n",
       "             mlogp      beta    sebeta    af_alt  af_alt_cases  \\\n",
       "0         0.972006 -0.114822  0.071168  0.005846      0.005683   \n",
       "1         0.207528 -0.021548  0.043470  0.013501      0.013448   \n",
       "2         0.065689 -0.023716  0.134105  0.001112      0.001117   \n",
       "3         0.492355 -0.215787  0.217818  0.000563      0.000556   \n",
       "4         0.116779  0.021523  0.071757  0.004192      0.004207   \n",
       "...            ...       ...       ...       ...           ...   \n",
       "20170229  0.360906  0.004066  0.005215  0.291210      0.291159   \n",
       "20170230  0.252027  0.025926  0.044450  0.003263      0.003298   \n",
       "20170231  2.117900 -0.043135  0.016165  0.024340      0.023984   \n",
       "20170232  0.795637  0.007715  0.005492  0.245151      0.245738   \n",
       "20170233  0.825873  0.007914  0.005489  0.245648      0.246278   \n",
       "\n",
       "          af_alt_controls  \n",
       "0                0.005914  \n",
       "1                0.013524  \n",
       "2                0.001109  \n",
       "3                0.000566  \n",
       "4                0.004186  \n",
       "...                   ...  \n",
       "20170229         0.291231  \n",
       "20170230         0.003248  \n",
       "20170231         0.024489  \n",
       "20170232         0.244904  \n",
       "20170233         0.245383  \n",
       "\n",
       "[20170234 rows x 13 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gwas"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "dc659323-8943-4e41-9f41-5cd0c8e2c17e",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>trait</th>\n",
       "      <th>region</th>\n",
       "      <th>v</th>\n",
       "      <th>cs</th>\n",
       "      <th>cs_specific_prob</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>position</th>\n",
       "      <th>allele1</th>\n",
       "      <th>allele2</th>\n",
       "      <th>maf</th>\n",
       "      <th>beta</th>\n",
       "      <th>p</th>\n",
       "      <th>se</th>\n",
       "      <th>most_severe</th>\n",
       "      <th>gene_most_severe</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>I9_HYPTENS</td>\n",
       "      <td>chr1:5120847-8120847</td>\n",
       "      <td>1:6584663:G:A</td>\n",
       "      <td>1</td>\n",
       "      <td>0.043208</td>\n",
       "      <td>chr1</td>\n",
       "      <td>6584663</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.281248</td>\n",
       "      <td>-0.038346</td>\n",
       "      <td>2.351200e-09</td>\n",
       "      <td>0.006421</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>ZBTB48</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>I9_HYPTENS</td>\n",
       "      <td>chr1:5120847-8120847</td>\n",
       "      <td>1:6585807:G:A</td>\n",
       "      <td>1</td>\n",
       "      <td>0.044924</td>\n",
       "      <td>chr1</td>\n",
       "      <td>6585807</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.281268</td>\n",
       "      <td>-0.038387</td>\n",
       "      <td>2.255640e-09</td>\n",
       "      <td>0.006421</td>\n",
       "      <td>non_coding_transcript_exon_variant</td>\n",
       "      <td>ZBTB48</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>I9_HYPTENS</td>\n",
       "      <td>chr1:5120847-8120847</td>\n",
       "      <td>1:6594993:G:A</td>\n",
       "      <td>1</td>\n",
       "      <td>0.057924</td>\n",
       "      <td>chr1</td>\n",
       "      <td>6594993</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.280470</td>\n",
       "      <td>-0.038723</td>\n",
       "      <td>1.677950e-09</td>\n",
       "      <td>0.006426</td>\n",
       "      <td>3_prime_UTR_variant</td>\n",
       "      <td>KLHL21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>I9_HYPTENS</td>\n",
       "      <td>chr1:5120847-8120847</td>\n",
       "      <td>1:6599445:G:A</td>\n",
       "      <td>1</td>\n",
       "      <td>0.078622</td>\n",
       "      <td>chr1</td>\n",
       "      <td>6599445</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>0.281519</td>\n",
       "      <td>-0.039011</td>\n",
       "      <td>1.202020e-09</td>\n",
       "      <td>0.006416</td>\n",
       "      <td>synonymous_variant</td>\n",
       "      <td>KLHL21</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>I9_HYPTENS</td>\n",
       "      <td>chr1:5120847-8120847</td>\n",
       "      <td>1:6600289:A:G</td>\n",
       "      <td>1</td>\n",
       "      <td>0.075465</td>\n",
       "      <td>chr1</td>\n",
       "      <td>6600289</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>0.281502</td>\n",
       "      <td>-0.038967</td>\n",
       "      <td>1.255360e-09</td>\n",
       "      <td>0.006416</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>KLHL21</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        trait                region              v  cs  cs_specific_prob  \\\n",
       "0  I9_HYPTENS  chr1:5120847-8120847  1:6584663:G:A   1          0.043208   \n",
       "1  I9_HYPTENS  chr1:5120847-8120847  1:6585807:G:A   1          0.044924   \n",
       "2  I9_HYPTENS  chr1:5120847-8120847  1:6594993:G:A   1          0.057924   \n",
       "3  I9_HYPTENS  chr1:5120847-8120847  1:6599445:G:A   1          0.078622   \n",
       "4  I9_HYPTENS  chr1:5120847-8120847  1:6600289:A:G   1          0.075465   \n",
       "\n",
       "  chromosome  position allele1 allele2       maf      beta             p  \\\n",
       "0       chr1   6584663       G       A  0.281248 -0.038346  2.351200e-09   \n",
       "1       chr1   6585807       G       A  0.281268 -0.038387  2.255640e-09   \n",
       "2       chr1   6594993       G       A  0.280470 -0.038723  1.677950e-09   \n",
       "3       chr1   6599445       G       A  0.281519 -0.039011  1.202020e-09   \n",
       "4       chr1   6600289       A       G  0.281502 -0.038967  1.255360e-09   \n",
       "\n",
       "         se                         most_severe gene_most_severe  \n",
       "0  0.006421                      intron_variant           ZBTB48  \n",
       "1  0.006421  non_coding_transcript_exon_variant           ZBTB48  \n",
       "2  0.006426                 3_prime_UTR_variant           KLHL21  \n",
       "3  0.006416                  synonymous_variant           KLHL21  \n",
       "4  0.006416                      intron_variant           KLHL21  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "finemap.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "04130f65-8780-46bf-acda-e559a1d77b58",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>trait</th>\n",
       "      <th>region</th>\n",
       "      <th>cs</th>\n",
       "      <th>cs_log10bf</th>\n",
       "      <th>cs_avg_r2</th>\n",
       "      <th>cs_min_r2</th>\n",
       "      <th>low_purity</th>\n",
       "      <th>cs_size</th>\n",
       "      <th>good_cs</th>\n",
       "      <th>cs_id</th>\n",
       "      <th>v</th>\n",
       "      <th>rsid</th>\n",
       "      <th>p</th>\n",
       "      <th>beta</th>\n",
       "      <th>sd</th>\n",
       "      <th>prob</th>\n",
       "      <th>cs_specific_prob</th>\n",
       "      <th>most_severe</th>\n",
       "      <th>gene_most_severe</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>I9_HYPTENS</td>\n",
       "      <td>chr11:365986-3365986</td>\n",
       "      <td>3</td>\n",
       "      <td>1.476225</td>\n",
       "      <td>0.317892</td>\n",
       "      <td>0.030052</td>\n",
       "      <td>True</td>\n",
       "      <td>11</td>\n",
       "      <td>False</td>\n",
       "      <td>chr11:365986-3365986_3</td>\n",
       "      <td>11:1899025:T:G</td>\n",
       "      <td>chr11_1899025_T_G</td>\n",
       "      <td>2.560350e-19</td>\n",
       "      <td>0.061645</td>\n",
       "      <td>0.005078</td>\n",
       "      <td>0.030545</td>\n",
       "      <td>0.021575</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>LINC01150</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>I9_HYPTENS</td>\n",
       "      <td>chr1:54982752-57982752</td>\n",
       "      <td>2</td>\n",
       "      <td>1.688606</td>\n",
       "      <td>0.267826</td>\n",
       "      <td>0.007267</td>\n",
       "      <td>True</td>\n",
       "      <td>61</td>\n",
       "      <td>False</td>\n",
       "      <td>chr1:54982752-57982752_2</td>\n",
       "      <td>1:57780626:A:G</td>\n",
       "      <td>chr1_57780626_A_G</td>\n",
       "      <td>4.735000e-08</td>\n",
       "      <td>-0.031565</td>\n",
       "      <td>0.011374</td>\n",
       "      <td>0.169756</td>\n",
       "      <td>0.169537</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>DAB1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>I9_HYPTENS</td>\n",
       "      <td>chr1:54982752-57982752</td>\n",
       "      <td>1</td>\n",
       "      <td>5.047026</td>\n",
       "      <td>0.882801</td>\n",
       "      <td>0.730739</td>\n",
       "      <td>False</td>\n",
       "      <td>15</td>\n",
       "      <td>True</td>\n",
       "      <td>chr1:54982752-57982752_1</td>\n",
       "      <td>1:56482752:T:C</td>\n",
       "      <td>chr1_56482752_T_C</td>\n",
       "      <td>1.398940e-11</td>\n",
       "      <td>0.040267</td>\n",
       "      <td>0.012328</td>\n",
       "      <td>0.113580</td>\n",
       "      <td>0.113364</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>AC119674.2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>I9_HYPTENS</td>\n",
       "      <td>chr16:80069630-83069630</td>\n",
       "      <td>2</td>\n",
       "      <td>1.799040</td>\n",
       "      <td>0.364302</td>\n",
       "      <td>0.048222</td>\n",
       "      <td>True</td>\n",
       "      <td>38</td>\n",
       "      <td>False</td>\n",
       "      <td>chr16:80069630-83069630_2</td>\n",
       "      <td>16:81476866:C:G</td>\n",
       "      <td>chr16_81476866_C_G</td>\n",
       "      <td>4.057890e-10</td>\n",
       "      <td>-0.041016</td>\n",
       "      <td>0.007396</td>\n",
       "      <td>0.051703</td>\n",
       "      <td>0.050699</td>\n",
       "      <td>intron_variant</td>\n",
       "      <td>CMIP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>I9_HYPTENS</td>\n",
       "      <td>chr13:20244714-23244714</td>\n",
       "      <td>1</td>\n",
       "      <td>5.553981</td>\n",
       "      <td>0.932232</td>\n",
       "      <td>0.871865</td>\n",
       "      <td>False</td>\n",
       "      <td>14</td>\n",
       "      <td>True</td>\n",
       "      <td>chr13:20244714-23244714_1</td>\n",
       "      <td>13:21744714:A:G</td>\n",
       "      <td>chr13_21744714_A_G</td>\n",
       "      <td>1.170040e-11</td>\n",
       "      <td>-0.040128</td>\n",
       "      <td>0.011729</td>\n",
       "      <td>0.096990</td>\n",
       "      <td>0.096802</td>\n",
       "      <td>intergenic_variant</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        trait                   region  cs  cs_log10bf  cs_avg_r2  cs_min_r2  \\\n",
       "0  I9_HYPTENS     chr11:365986-3365986   3    1.476225   0.317892   0.030052   \n",
       "1  I9_HYPTENS   chr1:54982752-57982752   2    1.688606   0.267826   0.007267   \n",
       "2  I9_HYPTENS   chr1:54982752-57982752   1    5.047026   0.882801   0.730739   \n",
       "3  I9_HYPTENS  chr16:80069630-83069630   2    1.799040   0.364302   0.048222   \n",
       "4  I9_HYPTENS  chr13:20244714-23244714   1    5.553981   0.932232   0.871865   \n",
       "\n",
       "   low_purity  cs_size  good_cs                      cs_id                v  \\\n",
       "0        True       11    False     chr11:365986-3365986_3   11:1899025:T:G   \n",
       "1        True       61    False   chr1:54982752-57982752_2   1:57780626:A:G   \n",
       "2       False       15     True   chr1:54982752-57982752_1   1:56482752:T:C   \n",
       "3        True       38    False  chr16:80069630-83069630_2  16:81476866:C:G   \n",
       "4       False       14     True  chr13:20244714-23244714_1  13:21744714:A:G   \n",
       "\n",
       "                 rsid             p      beta        sd      prob  \\\n",
       "0   chr11_1899025_T_G  2.560350e-19  0.061645  0.005078  0.030545   \n",
       "1   chr1_57780626_A_G  4.735000e-08 -0.031565  0.011374  0.169756   \n",
       "2   chr1_56482752_T_C  1.398940e-11  0.040267  0.012328  0.113580   \n",
       "3  chr16_81476866_C_G  4.057890e-10 -0.041016  0.007396  0.051703   \n",
       "4  chr13_21744714_A_G  1.170040e-11 -0.040128  0.011729  0.096990   \n",
       "\n",
       "   cs_specific_prob         most_severe gene_most_severe  \n",
       "0          0.021575      intron_variant        LINC01150  \n",
       "1          0.169537      intron_variant             DAB1  \n",
       "2          0.113364      intron_variant       AC119674.2  \n",
       "3          0.050699      intron_variant             CMIP  \n",
       "4          0.096802  intergenic_variant              NaN  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "precausal.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "8c5681a2-85f4-4fd3-a3b9-de8824651d5c",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>association_info.ancestry</th>\n",
       "      <th>association_info.doi</th>\n",
       "      <th>association_info.gwas_catalog_id</th>\n",
       "      <th>association_info.neg_log_pval</th>\n",
       "      <th>association_info.otg_id</th>\n",
       "      <th>association_info.pubmed_id</th>\n",
       "      <th>association_info.url</th>\n",
       "      <th>gold_standard_info.evidence.class</th>\n",
       "      <th>gold_standard_info.evidence.confidence</th>\n",
       "      <th>gold_standard_info.evidence.curated_by</th>\n",
       "      <th>...</th>\n",
       "      <th>sentinel_variant.alleles.alternative</th>\n",
       "      <th>sentinel_variant.alleles.reference</th>\n",
       "      <th>sentinel_variant.locus_GRCh37.chromosome</th>\n",
       "      <th>sentinel_variant.locus_GRCh37.position</th>\n",
       "      <th>sentinel_variant.locus_GRCh38.chromosome</th>\n",
       "      <th>sentinel_variant.locus_GRCh38.position</th>\n",
       "      <th>sentinel_variant.rsid</th>\n",
       "      <th>trait_info.ontology</th>\n",
       "      <th>trait_info.reported_trait_name</th>\n",
       "      <th>trait_info.standard_trait_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>European=408343</td>\n",
       "      <td>10.1038/s41588-018-0184-y</td>\n",
       "      <td>NaN</td>\n",
       "      <td>14.397940</td>\n",
       "      <td>SAIGE_401</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>drug</td>\n",
       "      <td>High</td>\n",
       "      <td>Ed Mountjoy</td>\n",
       "      <td>...</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>10</td>\n",
       "      <td>115805056</td>\n",
       "      <td>10</td>\n",
       "      <td>114045297</td>\n",
       "      <td>NaN</td>\n",
       "      <td>EFO_0000537</td>\n",
       "      <td>Hypertension</td>\n",
       "      <td>hypertension</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>European=408343</td>\n",
       "      <td>10.1038/s41588-018-0184-y</td>\n",
       "      <td>NaN</td>\n",
       "      <td>15.000000</td>\n",
       "      <td>SAIGE_401</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>drug</td>\n",
       "      <td>Low</td>\n",
       "      <td>Ed Mountjoy</td>\n",
       "      <td>...</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>11</td>\n",
       "      <td>1873232</td>\n",
       "      <td>11</td>\n",
       "      <td>1852002</td>\n",
       "      <td>NaN</td>\n",
       "      <td>EFO_0000537</td>\n",
       "      <td>Hypertension</td>\n",
       "      <td>hypertension</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>European=408343</td>\n",
       "      <td>10.1038/s41588-018-0184-y</td>\n",
       "      <td>NaN</td>\n",
       "      <td>19.619789</td>\n",
       "      <td>SAIGE_401</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>drug</td>\n",
       "      <td>Low</td>\n",
       "      <td>Ed Mountjoy</td>\n",
       "      <td>...</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>11</td>\n",
       "      <td>1888614</td>\n",
       "      <td>11</td>\n",
       "      <td>1867384</td>\n",
       "      <td>NaN</td>\n",
       "      <td>EFO_0000537</td>\n",
       "      <td>Hypertension</td>\n",
       "      <td>hypertension</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>European=69395</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GCST001238</td>\n",
       "      <td>10.301030</td>\n",
       "      <td>GCST001238</td>\n",
       "      <td>21909115.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>drug</td>\n",
       "      <td>Low</td>\n",
       "      <td>Ed Mountjoy</td>\n",
       "      <td>...</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>11</td>\n",
       "      <td>100593538</td>\n",
       "      <td>11</td>\n",
       "      <td>100722807</td>\n",
       "      <td>NaN</td>\n",
       "      <td>EFO_0000537</td>\n",
       "      <td>Hypertension</td>\n",
       "      <td>hypertension</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>European=408343</td>\n",
       "      <td>10.1038/s41588-018-0184-y</td>\n",
       "      <td>NaN</td>\n",
       "      <td>20.274905</td>\n",
       "      <td>SAIGE_401</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>drug</td>\n",
       "      <td>Low</td>\n",
       "      <td>Ed Mountjoy</td>\n",
       "      <td>...</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>11</td>\n",
       "      <td>100610546</td>\n",
       "      <td>11</td>\n",
       "      <td>100739815</td>\n",
       "      <td>NaN</td>\n",
       "      <td>EFO_0000537</td>\n",
       "      <td>Hypertension</td>\n",
       "      <td>hypertension</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 31 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  association_info.ancestry       association_info.doi  \\\n",
       "0           European=408343  10.1038/s41588-018-0184-y   \n",
       "1           European=408343  10.1038/s41588-018-0184-y   \n",
       "2           European=408343  10.1038/s41588-018-0184-y   \n",
       "3            European=69395                        NaN   \n",
       "4           European=408343  10.1038/s41588-018-0184-y   \n",
       "\n",
       "  association_info.gwas_catalog_id  association_info.neg_log_pval  \\\n",
       "0                              NaN                      14.397940   \n",
       "1                              NaN                      15.000000   \n",
       "2                              NaN                      19.619789   \n",
       "3                       GCST001238                      10.301030   \n",
       "4                              NaN                      20.274905   \n",
       "\n",
       "  association_info.otg_id  association_info.pubmed_id association_info.url  \\\n",
       "0               SAIGE_401                         NaN                  NaN   \n",
       "1               SAIGE_401                         NaN                  NaN   \n",
       "2               SAIGE_401                         NaN                  NaN   \n",
       "3              GCST001238                  21909115.0                  NaN   \n",
       "4               SAIGE_401                         NaN                  NaN   \n",
       "\n",
       "  gold_standard_info.evidence.class gold_standard_info.evidence.confidence  \\\n",
       "0                              drug                                   High   \n",
       "1                              drug                                    Low   \n",
       "2                              drug                                    Low   \n",
       "3                              drug                                    Low   \n",
       "4                              drug                                    Low   \n",
       "\n",
       "  gold_standard_info.evidence.curated_by  ...  \\\n",
       "0                            Ed Mountjoy  ...   \n",
       "1                            Ed Mountjoy  ...   \n",
       "2                            Ed Mountjoy  ...   \n",
       "3                            Ed Mountjoy  ...   \n",
       "4                            Ed Mountjoy  ...   \n",
       "\n",
       "  sentinel_variant.alleles.alternative  sentinel_variant.alleles.reference  \\\n",
       "0                                    C                                   G   \n",
       "1                                    T                                   A   \n",
       "2                                    T                                   C   \n",
       "3                                    C                                   G   \n",
       "4                                    C                                   T   \n",
       "\n",
       "  sentinel_variant.locus_GRCh37.chromosome  \\\n",
       "0                                       10   \n",
       "1                                       11   \n",
       "2                                       11   \n",
       "3                                       11   \n",
       "4                                       11   \n",
       "\n",
       "  sentinel_variant.locus_GRCh37.position  \\\n",
       "0                              115805056   \n",
       "1                                1873232   \n",
       "2                                1888614   \n",
       "3                              100593538   \n",
       "4                              100610546   \n",
       "\n",
       "  sentinel_variant.locus_GRCh38.chromosome  \\\n",
       "0                                       10   \n",
       "1                                       11   \n",
       "2                                       11   \n",
       "3                                       11   \n",
       "4                                       11   \n",
       "\n",
       "  sentinel_variant.locus_GRCh38.position sentinel_variant.rsid  \\\n",
       "0                              114045297                   NaN   \n",
       "1                                1852002                   NaN   \n",
       "2                                1867384                   NaN   \n",
       "3                              100722807                   NaN   \n",
       "4                              100739815                   NaN   \n",
       "\n",
       "  trait_info.ontology trait_info.reported_trait_name  \\\n",
       "0         EFO_0000537                   Hypertension   \n",
       "1         EFO_0000537                   Hypertension   \n",
       "2         EFO_0000537                   Hypertension   \n",
       "3         EFO_0000537                   Hypertension   \n",
       "4         EFO_0000537                   Hypertension   \n",
       "\n",
       "  trait_info.standard_trait_name  \n",
       "0                   hypertension  \n",
       "1                   hypertension  \n",
       "2                   hypertension  \n",
       "3                   hypertension  \n",
       "4                   hypertension  \n",
       "\n",
       "[5 rows x 31 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "causal.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "95020fad-f610-4ae3-876a-ebcaef2b2985",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "def explore_dataframe(dataframe, dataframe_name):\n",
    "    print(\"=== DataFrame Exploration: {} ===\".format(dataframe_name))\n",
    "    print(\"Number of Rows: {}\".format(dataframe.shape[0]))\n",
    "    print(\"Number of Columns: {}\".format(dataframe.shape[1]))\n",
    "    print(\"Column Names: {}\".format(\", \".join(dataframe.columns)))\n",
    "    print(\"\\nData Types of Columns:\")\n",
    "    print(dataframe.dtypes)\n",
    "    print(\"\\nNull Value Counts:\")\n",
    "    print(dataframe.isnull().sum())\n",
    "    print(\"\\nSummary Statistics:\")\n",
    "    print(dataframe.describe())\n",
    "    print(\"=== End of DataFrame Exploration: {} ===\\n\".format(dataframe_name))\n",
    "    \n",
    "#explore_dataframe(gwas, \"gwas\")\n",
    "#explore_dataframe(causal, \"causal\")\n",
    "#explore_dataframe(finemap, \"finemap\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c46b7243-adf9-42ac-9026-c5ac411159b6",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Data manipulation"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e8573bef-1cda-4e57-b3fa-f2b40333620a",
   "metadata": {
    "tags": []
   },
   "source": [
    "### Adjust `chromosome` in `finemap`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "825d60e9-ce09-43d2-8b4d-e5f47a942e90",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Extract number from 'chromosome' and replace 'X' with '23'\n",
    "finemap['chromosome'] = finemap['chromosome'].str.extract('(\\d+|X)', expand=False).replace('X', '23')\n",
    "\n",
    "# Convert 'chromosome' column to 'int64'\n",
    "finemap['chromosome'] = finemap['chromosome'].astype('int64')\n",
    "\n",
    "# Assertions to verify the data manipulations\n",
    "assert finemap['chromosome'].dtype == 'int64'  \n",
    "assert finemap['chromosome'].isin(range(1, 24)).all()  "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "337a14ab-c09e-42ae-b737-40f95d2080da",
   "metadata": {
    "tags": []
   },
   "source": [
    "### Adjust `v` in `finemap`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "a1c693d8-0ed6-4529-9c3c-3588f5034ce4",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Replace 'X' with '23' in 'v' column of finemap\n",
    "finemap['v'] = finemap['v'].str.replace(r'(^X:)', '23:', regex=True)\n",
    "\n",
    "# Assert 'X' is not in 'v' column anymore\n",
    "assert 'X' not in finemap['v']"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c30d6716-fce0-4ca7-9f2b-0070bda37dce",
   "metadata": {
    "tags": []
   },
   "source": [
    "### Create `finemapped` in `gwas`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "f90a6277-747b-4cc6-994c-06e772128d65",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Assertions passed successfully.\n"
     ]
    }
   ],
   "source": [
    "# Create the 'id' column in the 'gwas' DataFrame\n",
    "gwas['id'] = gwas['#chrom'].astype(str) + ':' + gwas['pos'].astype(str) + ':' + gwas['ref'] + ':' + gwas['alt']\n",
    "\n",
    "# Create a set for faster lookup\n",
    "finemap_set = set(finemap['v'].values)\n",
    "\n",
    "# Use the set for lookup\n",
    "gwas['finemapped'] = gwas['id'].apply(lambda x: 1 if x in finemap_set else 0)\n",
    "\n",
    "# Count the number of 1s in the 'finemapped' column\n",
    "count_ones = gwas['finemapped'].sum()\n",
    "\n",
    "# Perform assertions to validate the results\n",
    "assert len(gwas) == len(gwas['id']) == len(gwas['finemapped']), \"Lengths do not match.\"\n",
    "assert count_ones <= len(gwas), \"Invalid count of 1s.\"\n",
    "\n",
    "print(\"Assertions passed successfully.\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9069885a-cfb2-42cd-8c72-c0f4ee5be92b",
   "metadata": {
    "tags": []
   },
   "source": [
    "### Create `precausal` in `gwas`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "5a9b986d-225e-4a16-acc6-d9ea9e283904",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "288\n"
     ]
    }
   ],
   "source": [
    "precausal_snp_set = set(precausal['v'])\n",
    "\n",
    "gwas['precausal'] = gwas['id'].apply(lambda x: 1 if x in precausal_snp_set else 0)\n",
    "print(gwas['precausal'].sum())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9b39302a-cdfc-4f03-9ca7-e7c7369895a0",
   "metadata": {
    "tags": []
   },
   "source": [
    "### Create `causal` in `gwas`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "bacd3fb6-07c7-4d01-a81a-3a0fa90757ad",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "33\n"
     ]
    }
   ],
   "source": [
    "# create a combined column in both dataframes\n",
    "causal['combined'] = causal['sentinel_variant.locus_GRCh38.chromosome'].astype(str) + '_' + causal['sentinel_variant.locus_GRCh38.position'].astype(str)\n",
    "gwas['combined'] = gwas['#chrom'].astype(str) + '_' + gwas['pos'].astype(str)\n",
    "\n",
    "# create a set from the 'combined' column in causal\n",
    "causal_combined_set = set(causal['combined'])\n",
    "\n",
    "# create a new column 'causal' in gwas that checks if 'combined' is in causal_combined_set\n",
    "gwas['causal'] = gwas['combined'].apply(lambda x: 1 if x in causal_combined_set else 0)\n",
    "\n",
    "print(gwas['causal'].sum())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d5e503e8-1146-47d2-a25c-0f57905d4994",
   "metadata": {
    "tags": []
   },
   "source": [
    "### Extract `trait` from `finemap` to `gwas`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "159fbc16-3d9c-4e80-b8dd-271a2088b20d",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "unique_trait = finemap['trait'].unique()\n",
    "trait_string = unique_trait[0]\n",
    "gwas['trait'] = trait_string"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8a4694be-acb8-4d22-9053-d985fe51fef6",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Export csv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "87b67bf8-bfbe-49a3-9e45-381ed4dec535",
   "metadata": {},
   "outputs": [],
   "source": [
    "finemapped_gwas = gwas[(gwas['finemapped'] == 1) | (gwas['precausal'] == 1)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "d0efede1-d7ac-4bf8-9631-4cecc4e2ce46",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "288\n"
     ]
    }
   ],
   "source": [
    "print(finemapped_gwas['precausal'].sum())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "4440d869-be1f-4439-8bd5-428331f0769f",
   "metadata": {},
   "outputs": [],
   "source": [
    "finemapped_gwas = finemapped_gwas.drop(['causal', 'finemapped', 'combined'], axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "3e156f19-8d30-4c06-8de1-a520bceb939b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#chrom</th>\n",
       "      <th>pos</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>rsids</th>\n",
       "      <th>nearest_genes</th>\n",
       "      <th>pval</th>\n",
       "      <th>mlogp</th>\n",
       "      <th>beta</th>\n",
       "      <th>sebeta</th>\n",
       "      <th>af_alt</th>\n",
       "      <th>af_alt_cases</th>\n",
       "      <th>af_alt_controls</th>\n",
       "      <th>id</th>\n",
       "      <th>precausal</th>\n",
       "      <th>trait</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>23344</th>\n",
       "      <td>1</td>\n",
       "      <td>3412095</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>rs2493292</td>\n",
       "      <td>PRDM16</td>\n",
       "      <td>8.673610e-11</td>\n",
       "      <td>10.06180</td>\n",
       "      <td>0.059679</td>\n",
       "      <td>0.009198</td>\n",
       "      <td>0.109426</td>\n",
       "      <td>0.112670</td>\n",
       "      <td>0.108063</td>\n",
       "      <td>1:3412095:C:T</td>\n",
       "      <td>1</td>\n",
       "      <td>I9_HYPTENS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50476</th>\n",
       "      <td>1</td>\n",
       "      <td>6584663</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>rs731024</td>\n",
       "      <td>ZBTB48</td>\n",
       "      <td>2.351200e-09</td>\n",
       "      <td>8.62871</td>\n",
       "      <td>-0.038346</td>\n",
       "      <td>0.006421</td>\n",
       "      <td>0.281248</td>\n",
       "      <td>0.277440</td>\n",
       "      <td>0.282847</td>\n",
       "      <td>1:6584663:G:A</td>\n",
       "      <td>0</td>\n",
       "      <td>I9_HYPTENS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50484</th>\n",
       "      <td>1</td>\n",
       "      <td>6585807</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>rs6682920</td>\n",
       "      <td>ZBTB48</td>\n",
       "      <td>2.255640e-09</td>\n",
       "      <td>8.64673</td>\n",
       "      <td>-0.038387</td>\n",
       "      <td>0.006421</td>\n",
       "      <td>0.281268</td>\n",
       "      <td>0.277461</td>\n",
       "      <td>0.282867</td>\n",
       "      <td>1:6585807:G:A</td>\n",
       "      <td>0</td>\n",
       "      <td>I9_HYPTENS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50547</th>\n",
       "      <td>1</td>\n",
       "      <td>6594993</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>rs2235566</td>\n",
       "      <td>KLHL21</td>\n",
       "      <td>1.677950e-09</td>\n",
       "      <td>8.77522</td>\n",
       "      <td>-0.038723</td>\n",
       "      <td>0.006426</td>\n",
       "      <td>0.280470</td>\n",
       "      <td>0.276618</td>\n",
       "      <td>0.282088</td>\n",
       "      <td>1:6594993:G:A</td>\n",
       "      <td>0</td>\n",
       "      <td>I9_HYPTENS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50597</th>\n",
       "      <td>1</td>\n",
       "      <td>6599445</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>rs2232460</td>\n",
       "      <td>KLHL21</td>\n",
       "      <td>1.202020e-09</td>\n",
       "      <td>8.92009</td>\n",
       "      <td>-0.039011</td>\n",
       "      <td>0.006416</td>\n",
       "      <td>0.281519</td>\n",
       "      <td>0.277570</td>\n",
       "      <td>0.283177</td>\n",
       "      <td>1:6599445:G:A</td>\n",
       "      <td>0</td>\n",
       "      <td>I9_HYPTENS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19925078</th>\n",
       "      <td>23</td>\n",
       "      <td>103566142</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.266050e-08</td>\n",
       "      <td>7.64473</td>\n",
       "      <td>0.090374</td>\n",
       "      <td>0.016166</td>\n",
       "      <td>0.021548</td>\n",
       "      <td>0.023120</td>\n",
       "      <td>0.020887</td>\n",
       "      <td>23:103566142:C:T</td>\n",
       "      <td>0</td>\n",
       "      <td>I9_HYPTENS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19925210</th>\n",
       "      <td>23</td>\n",
       "      <td>103613602</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.810740e-08</td>\n",
       "      <td>7.55118</td>\n",
       "      <td>0.089673</td>\n",
       "      <td>0.016149</td>\n",
       "      <td>0.021587</td>\n",
       "      <td>0.023138</td>\n",
       "      <td>0.020935</td>\n",
       "      <td>23:103613602:G:A</td>\n",
       "      <td>0</td>\n",
       "      <td>I9_HYPTENS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19925235</th>\n",
       "      <td>23</td>\n",
       "      <td>103622280</td>\n",
       "      <td>GTC</td>\n",
       "      <td>G</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.325630e-08</td>\n",
       "      <td>7.63346</td>\n",
       "      <td>0.090112</td>\n",
       "      <td>0.016132</td>\n",
       "      <td>0.021632</td>\n",
       "      <td>0.023192</td>\n",
       "      <td>0.020976</td>\n",
       "      <td>23:103622280:GTC:G</td>\n",
       "      <td>0</td>\n",
       "      <td>I9_HYPTENS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19925467</th>\n",
       "      <td>23</td>\n",
       "      <td>103686440</td>\n",
       "      <td>CT</td>\n",
       "      <td>C</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>8.519420e-08</td>\n",
       "      <td>7.06959</td>\n",
       "      <td>0.098822</td>\n",
       "      <td>0.018451</td>\n",
       "      <td>0.016685</td>\n",
       "      <td>0.017929</td>\n",
       "      <td>0.016162</td>\n",
       "      <td>23:103686440:CT:C</td>\n",
       "      <td>0</td>\n",
       "      <td>I9_HYPTENS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19925678</th>\n",
       "      <td>23</td>\n",
       "      <td>103737629</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3.609520e-08</td>\n",
       "      <td>7.44255</td>\n",
       "      <td>0.089289</td>\n",
       "      <td>0.016208</td>\n",
       "      <td>0.021694</td>\n",
       "      <td>0.023196</td>\n",
       "      <td>0.021063</td>\n",
       "      <td>23:103737629:A:G</td>\n",
       "      <td>0</td>\n",
       "      <td>I9_HYPTENS</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>7497 rows × 16 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          #chrom        pos  ref alt      rsids nearest_genes          pval  \\\n",
       "23344          1    3412095    C   T  rs2493292        PRDM16  8.673610e-11   \n",
       "50476          1    6584663    G   A   rs731024        ZBTB48  2.351200e-09   \n",
       "50484          1    6585807    G   A  rs6682920        ZBTB48  2.255640e-09   \n",
       "50547          1    6594993    G   A  rs2235566        KLHL21  1.677950e-09   \n",
       "50597          1    6599445    G   A  rs2232460        KLHL21  1.202020e-09   \n",
       "...          ...        ...  ...  ..        ...           ...           ...   \n",
       "19925078      23  103566142    C   T        NaN           NaN  2.266050e-08   \n",
       "19925210      23  103613602    G   A        NaN           NaN  2.810740e-08   \n",
       "19925235      23  103622280  GTC   G        NaN           NaN  2.325630e-08   \n",
       "19925467      23  103686440   CT   C        NaN           NaN  8.519420e-08   \n",
       "19925678      23  103737629    A   G        NaN           NaN  3.609520e-08   \n",
       "\n",
       "             mlogp      beta    sebeta    af_alt  af_alt_cases  \\\n",
       "23344     10.06180  0.059679  0.009198  0.109426      0.112670   \n",
       "50476      8.62871 -0.038346  0.006421  0.281248      0.277440   \n",
       "50484      8.64673 -0.038387  0.006421  0.281268      0.277461   \n",
       "50547      8.77522 -0.038723  0.006426  0.280470      0.276618   \n",
       "50597      8.92009 -0.039011  0.006416  0.281519      0.277570   \n",
       "...            ...       ...       ...       ...           ...   \n",
       "19925078   7.64473  0.090374  0.016166  0.021548      0.023120   \n",
       "19925210   7.55118  0.089673  0.016149  0.021587      0.023138   \n",
       "19925235   7.63346  0.090112  0.016132  0.021632      0.023192   \n",
       "19925467   7.06959  0.098822  0.018451  0.016685      0.017929   \n",
       "19925678   7.44255  0.089289  0.016208  0.021694      0.023196   \n",
       "\n",
       "          af_alt_controls                  id  precausal       trait  \n",
       "23344            0.108063       1:3412095:C:T          1  I9_HYPTENS  \n",
       "50476            0.282847       1:6584663:G:A          0  I9_HYPTENS  \n",
       "50484            0.282867       1:6585807:G:A          0  I9_HYPTENS  \n",
       "50547            0.282088       1:6594993:G:A          0  I9_HYPTENS  \n",
       "50597            0.283177       1:6599445:G:A          0  I9_HYPTENS  \n",
       "...                   ...                 ...        ...         ...  \n",
       "19925078         0.020887    23:103566142:C:T          0  I9_HYPTENS  \n",
       "19925210         0.020935    23:103613602:G:A          0  I9_HYPTENS  \n",
       "19925235         0.020976  23:103622280:GTC:G          0  I9_HYPTENS  \n",
       "19925467         0.016162   23:103686440:CT:C          0  I9_HYPTENS  \n",
       "19925678         0.021063    23:103737629:A:G          0  I9_HYPTENS  \n",
       "\n",
       "[7497 rows x 16 columns]"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "finemapped_gwas"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "ac23a3cb-009e-4d63-83bf-d4639613ec1f",
   "metadata": {},
   "outputs": [],
   "source": [
    "#finemapped_gwas.to_csv('hyp-finemapped-precausal.csv', index=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
